### PR TITLE
Correctly check for overflow in add, mul and pow

### PR DESCRIPTION
**Pull request by SwayStar123**
## Description
Adds flag checks for overflow in core lib, properly cap values if overflow is enabled

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
   - [ ] If my change requires substantial documentation changes, I have [requested support from the DevRel team](https://github.com/FuelLabs/devrel-requests/issues/new/choose)
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by Dentosal** [comment]
This is not a vm issue, the instruction that's supposed to revert is getting optimized away. Likely the compiler doesn't consider panic-on-invalid-input side effect. Not sure if the compiler behavior is expected or not.

## Investigation

if the function looks like this:
```sway
#[test(should_revert)]
fn math_0th_root_fail() {
    let _res = asm(r1: 100, r2: 0, r3) {
        log flag r1 r2 zero;
        mroo r3 r1 r2;
        log flag r1 r2 r3;
        log one one one one;
        r3: u8
    };
}
```
Then the test passes as expected.

However, if we remove the middle log, i.e.:

```sway
#[test(should_revert)]
fn math_0th_root_fail() {
    let _res = asm(r1: 100, r2: 0, r3) {
        log flag r1 r2 zero;
        mroo r3 r1 r2;
        log one one one one;
        r3: u8
    };
}
```
Then the test fails.

Let's look at the bytecode. Compile with forc build --output-bin prog.bin --tests  (in test/src/in_language_tests/test_programs/math_inline_tests). Then we can find the relevant log instructions with forc parse-bytecode prog.bin | rg -C 1 '\bLOG \{'

For the first version, i.e. without logging r3 , the bytecode looks like this:
```
      18230   72920   MOVE { dst: 0x11, src: 0x0 }                                               1a 44 00 00
      18231   72924   LOG { a: 0xf, b: 0x10, c: 0x11, d: 0x0 }                                   33 3d 04 40
      18232   72928   LOG { a: 0x1, b: 0x1, c: 0x1, d: 0x1 }                                     33 04 10 41
      18233   72932   RET { value: 0x0 }                                                         24 00 00 00
```
However, when we introduce the log in the code in the second version, we get:
```
      18230   72920   MOVE { dst: 0x11, src: 0x0 }                                               1a 44 00 00
      18231   72924   LOG { a: 0xf, b: 0x10, c: 0x11, d: 0x0 }                                   33 3d 04 40
      18232   72928   MROO { dst: 0x12, lhs: 0x10, rhs: 0x11 }                                   18 49 04 40
      18233   72932   LOG { a: 0xf, b: 0x10, c: 0x11, d: 0x12 }                                  33 3d 04 52
      18234   72936   LOG { a: 0x1, b: 0x1, c: 0x1, d: 0x1 }                                     33 04 10 41
      18235   72940   RET { value: 0x0 }                                                         24 00 00 00
```
So it seems like the mroo is completely missing in the first version.

File: sway-lib-std/src/math.sw [hunk 1]
```diff
@@ -97,13 +121,28 @@ impl Power for u256 {
 
         while exp > 1 {
             if (exp & 1) == 1 {
-                acc = acc * base;
+                // acc = acc * base;
+                let res = u256_checked_mul(acc, base);
+                acc = match res {
+                    Option::None => return u256::zero(),
+                    Option::Some(val) => val,
```
Comment (bitzoic) [inline sway-lib-std/src/math.sw:128]:
```suggestion
                    Some(val) => val,
                    None => return u256::zero(),
```
nit

File: sway-lib-std/src/math.sw [hunk 2]
```diff
@@ -97,13 +121,28 @@ impl Power for u256 {
...
+                }
             }
             exp = exp >> 1;
-            base = base * base;
+            // base = base * base;
+            let res = u256_checked_mul(base, base);
+            base = match res {
+                Option::None => return u256::zero(),
+                Option::Some(val) => val,
```
Comment (bitzoic) [inline sway-lib-std/src/math.sw:136]:
```suggestion
                Some(val) => val,
                None => return u256::zero(),
```
nit

File: sway-lib-std/src/math.sw [hunk 3]
```diff
@@ -97,13 +121,28 @@ impl Power for u256 {
...
+            }
         }
 
-        acc * base
+        // acc * base
+        let res = u256_checked_mul(acc, base);
+        match res {
+            Option::None => u256::zero(),
+            Option::Some(val) => val,
```
Comment (bitzoic) [inline sway-lib-std/src/math.sw:144]:
```suggestion
            Some(val) => val,
            None => u256::zero(),
```

File: sway-lib-std/src/math.sw [hunk 4]
```diff
@@ -118,14 +157,21 @@ impl Power for u64 {
...
             exp r3 r1 r2;
             r3: u64
         };
-        // If panic on wrapping math is enabled, only then revert
-        if flags() & F_WRAPPING_DISABLE_MASK == 0 {
-            assert(res <= Self::max().as_u64());
+
+        if res > Self::max().as_u64() {
+            // If panic on wrapping math is enabled, only then revert
+            if flags() & F_WRAPPING_DISABLE_MASK == 0 {
```
Comment (bitzoic) [inline sway-lib-std/src/math.sw:167]: Isn't there a `panic_on_overflow_enabled()` function for the std-lib?

File: test/src/in_language_tests/test_programs/math_inline_tests/src/main.sw [hunk 5]
```diff
@@ -819,3 +821,223 @@ fn math_0th_root_fail() {
...
+}
+
+#[test]
+fn math_u8_overflow_add() {
+    let _ = disable_panic_on_overflow();
+
+    let a = u8::max();
+    let b = a + 1;
+
+    require(b == 0_u8, b);
```
Comment (bitzoic) [inline test/src/in_language_tests/test_programs/math_inline_tests/src/main.sw:902]:
```suggestion
    let _ = disable_panic_on_overflow();

    let a = u8::max();

    // Wrap test
    let b = a + 1;
    require(b == 0_u8, b);
    
    // Wraps plus one
    let c = a + 2;
    require(c == 1_u8, c);
    
    // Double
    let d = a + u8::max();
    require(c == 0_u8, c);
    
    // etc
```
Some more edge cases would be nice

Comment (bitzoic) [inline test/src/in_language_tests/test_programs/math_inline_tests/src/main.sw:902]:
```suggestion
    let _ = disable_panic_on_overflow();

    let a = u8::max();

    // Wrap test
    let b = a + 1;
    require(b == 0_u8, b);
    
    // Wraps plus one
    let c = a + 2;
    require(c == 1_u8, c);
    
    // Double
    let d = a + u8::max();
    require(c == 0_u8, c);
```
Some more edge cases would be nice

File: sway-lib-std/src/math.sw [hunk 6]
```diff
@@ -78,6 +80,28 @@ pub trait Power {
...
+    // .X....	indirect0	Is lhs operand ($rB) indirect or not
+    // X.....	indirect1	Is rhs operand ($rC) indirect or not
+    // As both operands are indirect, 110000 is used, which is 48 in decimal.
+    let of = asm(res: res, a: a, b: b) {
+        wqml res a b i48;
+        of: u64
+    };
+
+    if of != 0 {
+        return Option::None;
```
Comment (bitzoic) [inline sway-lib-std/src/math.sw:99]:
```suggestion
        return None;
```

File: sway-lib-std/src/math.sw [hunk 7]
```diff
@@ -78,6 +80,28 @@ pub trait Power {
...
+    }
+
+    Option::Some(res)
```
Comment (bitzoic) [inline sway-lib-std/src/math.sw:102]:
```suggestion
    Some(res)
```

File: sway-lib-core/src/ops.sw [hunk 8]
```diff
@@ -73,7 +79,11 @@ impl Add for u16 {
     fn add(self, other: Self) -> Self {
         let res = __add(self, other);
         if __gt(res, Self::max()) {
-            __revert(0)
+            if panic_on_overflow_enabled() {
+                __revert(0)
+            } else {
+                __mod(res, __add(Self::max(), 1))
```
Comment (bitzoic) [inline sway-lib-core/src/ops.sw:85]:
```suggestion
            } else {
                // overflow enabled
                // res % (Self::max() + 1)
                __mod(res, __add(Self::max(), 1))
```

File: sway-lib-core/src/ops.sw [hunk 9]
```diff
@@ -93,7 +103,14 @@ impl Add for u8 {
             input: u64
         };
         if __gt(res_u64, max_u8_u64) {
-            __revert(0)
+            if panic_on_overflow_enabled() {
+                __revert(0)
+            } else {
+                let res_u64 = __mod(res_u64, __add(max_u8_u64, 1));
```
Comment (bitzoic) [inline sway-lib-core/src/ops.sw:109]:
```suggestion
            } else {
                // overflow enabled
                // res % (Self::max() + 1)
                let res_u64 = __mod(res_u64, __add(max_u8_u64, 1));
```

File: sway-lib-core/src/ops.sw [hunk 10]
```diff
@@ -262,7 +289,14 @@ impl Multiply for u8 {
...
```
Comment (bitzoic) [inline sway-lib-core/src/ops.sw:295]:
```suggestion
            } else {
                // overflow enabled
                // res % (Self::max() + 1)
                let res_u64 = __mod(res_u64, __add(max_u8_u64, 1));
```

File: sway-lib-core/src/ops.sw [hunk 11]
```diff
@@ -1258,3 +1292,19 @@ pub fn ok_str_eq() {
     assert("" != "a");
     assert("a" != "b");
 }
+
+fn flags() -> u64 {
```
Comment (bitzoic) [inline sway-lib-core/src/ops.sw:1296]: Is this function needed?

File: sway-lib-std/src/math.sw [hunk 12]
```diff
@@ -97,13 +121,28 @@ impl Power for u256 {
 
         while exp > 1 {
             if (exp & 1) == 1 {
-                acc = acc * base;
+                // acc = acc * base;
+                let res = u256_checked_mul(acc, base);
+                acc = match res {
+                    Option::None => return u256::zero(),
```
Comment (bitzoic) [inline sway-lib-std/src/math.sw:127]: Why do you return `u256::zero()` here?

File: test/src/in_language_tests/test_programs/math_inline_tests/src/main.sw [hunk 13]
```diff
@@ -819,3 +821,223 @@ fn math_0th_root_fail() {
...
+}
+
+#[test]
+fn math_u256_overflow_add() {
+    let _ = disable_panic_on_overflow();
+
+    let a = u256::max();
+    let b = a + 1;
+
+    require(b == 0.as_u256(), b);
```
Comment (bitzoic) [inline test/src/in_language_tests/test_programs/math_inline_tests/src/main.sw:942]:
```suggestion
    require(b == u256::zero(), b);
```

File: sway-lib-core/src/ops.sw (hunk 11 above)
Comment (SwayStar123) [reply sway-lib-core/src/ops.sw:1296]: I suppose not at the very moment, but in the future if you want to add a panic_on_unsafe_math_enabled() function you would need it seperate

File: sway-lib-std/src/math.sw (hunk 4 above)
Comment (SwayStar123) [reply sway-lib-std/src/math.sw:167]: that pr isnt merged yet, if that gets in before this ill change this

File: sway-lib-std/src/math.sw (hunk 12 above)
Comment (SwayStar123) [reply sway-lib-std/src/math.sw]: Following the spec of u64. If u64 overflows on a pow operation it returns 0.

File: sway-lib-core/src/ops.sw [hunk 14]
```diff
@@ -1258,3 +1298,19 @@ pub fn ok_str_eq() {
...
+    asm() {
+        flag
+    }
+}
+
+fn panic_on_overflow_enabled() -> bool {
```
Comment (IGI-111) [inline sway-lib-core/src/ops.sw:1308]: This doesn't actuall panic and is just a predicate.

Call it `is_overflow_enabled` instead or include the panic in the procedure.

Comment (SwayStar123) [reply sway-lib-core/src/ops.sw:1308]: it is not "panic on (overflow enabled)", it is "(panic on overflow) enabled". I suppose it is a tad bit confusing but the flag  behavior is described as "panic on overflow" everywhere else now, and all other functions use it, so i believe this would be the correct name following the naming convention of declarative names ("enabled" over "is_on") that are followed, atleast in sway applications. Is this not the convention for sway?

Comment (SwayStar123) [reply sway-lib-core/src/ops.sw]: Ive renamed it in this pr, but theres another function already in the std lib now with the old name of panic_on_overflow_enabled, which needs to be changed in another PR now

**Review by bitzoic — APPROVED** [review]

**Review by IGI-111 — APPROVED** [review]
